overlord/snapstate/handlers.go

```diff
@@ -1166,29 +1241,47 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) (err error) {
 		return copyDataErr
 	}
 
-	action := triggeredMigration(newInfo.Base, opts)
-	var undo func()
+	var oldBase string
+	if oldInfo != nil {
+		oldBase = oldInfo.Base
+	}
 
-	switch action {
+	snapName := snapsup.InstanceName()
+	switch triggeredMigration(oldBase, newInfo.Base, opts) {
 	case hidden:
-		undo, err = m.doHiddenMigration(t, st, snapsup, newInfo)
+		if err := m.backend.HideSnapData(snapName); err != nil {
+			return err
+		}
+
+		snapsup.MigratedHidden = true
 	case revertHidden:
-		undo, err = m.revertHiddenMigration(t, st, snapsup)
+		if err := m.backend.UndoHideSnapData(snapName); err != nil {
+			return err
+		}
+
+		snapsup.UndidHiddenMigration = true
 	case home:
-		undo, err = m.doHomeMigration(t, st, snapsup, newInfo.Revision)
+		if err := m.backend.InitExposedSnapHome(snapName, newInfo.Revision); err != nil {
+			return err
+		}
+
+		snapsup.MigratedToExposedHome = true
 	case full:
-		undo, err = m.doFullMigration(t, st, snapsup, newInfo)
-	default:
-		// for now, don't support reverting the ~/Snap migration in a refresh
-		err = nil
-	}
+		if err := m.backend.HideSnapData(snapName); err != nil {
+			return err
+		}
 
-	// if task fails, undo any migration actions taken
-	defer func() {
-		if err != nil && undo != nil {
-			undo()
+		if err := m.backend.InitExposedSnapHome(snapName, newInfo.Revision); err != nil {
+			return err
 		}
-	}()
+
+		snapsup.MigratedHidden = true
+		snapsup.MigratedToExposedHome = true
+	}
+
+	st.Lock()
+	err = SetTaskSnapSetup(t, snapsup)
```

Comment:
I would try to use Lock,defer Unlock around SetTaskSnapSetup calls, as setting can panic